### PR TITLE
2020 10 05 redo config

### LINKED ISSUE #2086: AppConfig.getBaseConfig() does not seem to use application.conf correctly
Using the dlc oracle as a library, i've specified this `bitcoin-s` configuration in my applications `application.conf` file. 
```
bitcoin-s {
  network = regtest //testnet3, mainnet
  common {
    profile = "slick.jdbc.PostgresProfile$"
    db {
      url = "jdbc:postgresql://localhost:5432/oracle"
      driver = "org.postgresql.Driver"
      username = ""
      password = ""
      numThreads = 1 # I worry about async issues with signing and r values, so let's keep 1 thread
    }
  }

   oracle.profile = ${bitcoin-s.common.profile}
   oracle.db = ${bitcoin-s.common.db}
}
```

Here is the error message I am getting from scala playframework

>com.typesafe.config.ConfigException$UnresolvedSubstitution: reference.conf @ jar:file:/home/suredbits/.cache/coursier/v1/https/repo1.maven.org/maven2/com/typesafe/play/play-server_2.13/2.8.2/play-server_2.13-2.8.2.jar!/reference.conf: 90: Could not resolve substitution to a value: ${play.server.dir}

Which doesn't really make sense as i'm not messing around with play configuration at all.

Looking deep down the stack trace I see 

```
    com.typesafe.config.impl.SimpleConfig.resolve(SimpleConfig.java:41)
     org.bitcoins.db.AppConfig$.getBaseConfig(AppConfig.scala:229)
     org.bitcoins.db.AppConfig.baseConfig$lzycompute(AppConfig.scala:133)
     org.bitcoins.db.AppConfig.baseConfig(AppConfig.scala:132)
     org.bitcoins.db.AppConfig.config$lzycompute(AppConfig.scala:137)
     org.bitcoins.db.AppConfig.config(AppConfig.scala:136)
     org.bitcoins.db.AppConfig.chain$lzycompute(AppConfig.scala:120)
     org.bitcoins.db.AppConfig.chain(AppConfig.scala:119)
     org.bitcoins.db.AppConfig.network$lzycompute(AppConfig.scala:126)
     org.bitcoins.db.AppConfig.network(AppConfig.scala:126)
     org.bitcoins.db.AppConfig.<init>(AppConfig.scala:150)
     org.bitcoins.dlc.oracle.DLCOracleAppConfig.<init>(DLCOracleAppConfig.scala:17)
```

when i'm trying to initialize the `DLCOracleAppConfig`. 

I'm using a temporary datadir for now that does NOT contain a`bitcoin-s.conf`. So all configuration is in my `application.conf`. A quick glance at [`AppConfig.getBaseConfig()`](https://github.com/bitcoin-s/bitcoin-s/blob/856f88f0ddbb7adc437ecb1cdc6ebf3c7128cd76/db-commons/src/main/scala/org/bitcoins/db/AppConfig.scala#L175) seems to indicate we do some sort resolution between the too, but i'm not sure how this is applied. 

Ultimately the line i'm erroring out on is here: https://github.com/bitcoin-s/bitcoin-s/blob/856f88f0ddbb7adc437ecb1cdc6ebf3c7128cd76/db-commons/src/main/scala/org/bitcoins/db/AppConfig.scala#L229

### PR DESCRIPTION
**Pull request by Christewart**
This attempts to fix #2086 

Basically this moves us to using the top level config for resolving things (`bitcoin-s`) rather than diving into the `bitcoin-s` namespace. 

Here is what I do in this PR

1. When resolving things, use absolute paths rather than relative (`bitcoin-s.chain.filterBatchSize` rather than `chain.filterBatchSize`)
2. Make `AppConfig.config` protected so that everyone doesn't have access to it.
3. Make things `lazy` in the `AppConfig` so configuration resolution is not eager. This means we don't resolve the configuration too early when using bitcoin-s as a library (I think this is the root cause of #2086, still testing though).

### CODE REVIEW

File: project/build.properties [hunk 1]
```diff
@@ -1 +1 @@
-sbt.version=1.3.13
+sbt.version=1.4.0
```
Comment (benthecarman) [inline project/build.properties]: Is this needed?

Comment (Christewart) [reply project/build.properties]: No sorry testing out the new version :P

**Review by benthecarman — CHANGES_REQUESTED** [review]
Need to change the `BitcoindRpcAppConfig` to use absolute paths

**Comment by benthecarman** [comment]
Looks like this fixes #2124

**Comment by Christewart** [comment]
> Need to change the `BitcoindRpcAppConfig` to use absolute paths

It's concerning this didn't cause CI failures :raised_eyebrow:

File: dlc-oracle/src/main/scala/org/bitcoins/dlc/oracle/DLCOracleAppConfig.scala [hunk 2]
```diff
@@ -45,16 +45,17 @@ case class DLCOracleAppConfig(
...
+    for {
+      _ <- super.start()
+    } yield {
+      if (Files.notExists(datadir)) {
+        Files.createDirectories(datadir)
+      }
+      val numMigrations = {
+        migrate()
+      }
+      logger.info(s"Applied $numMigrations to the wallet project")
```
Comment (benthecarman) [inline dlc-oracle/src/main/scala/org/bitcoins/dlc/oracle/DLCOracleAppConfig.scala]: > Applied $numMigrations to the wallet project

this is not the wallet

File: wallet/src/main/scala/org/bitcoins/wallet/config/WalletAppConfig.scala [hunk 3]
```diff
@@ -82,33 +82,38 @@ case class WalletAppConfig(
...
+    config.getInt("bitcoin-s.wallet.requiredConfirmations")
 
   override def start(): Future[Unit] = {
-    logger.debug(s"Initializing wallet setup")
+    for {
+      _ <- super.start()
+      //moved out of main class body because we need lazy behavior when loading configs
+      _ = require(
+        requiredConfirmations >= 1,
+        s"requiredConfirmations cannot be less than 1, got: $requiredConfirmations")
```
Comment (benthecarman) [inline wallet/src/main/scala/org/bitcoins/wallet/config/WalletAppConfig.scala]: I think it'd be cleaner to ust change the definition of `requiredConfirmations` to
```scala

lazy val requiredConfirmations: Int = {
    confs = config.getInt("wallet.requiredConfirmations")
	require(
        confs >= 1,
        s"requiredConfirmations cannot be less than 1, got: $confs")
	confs
}

```

**Review by benthecarman — CHANGES_REQUESTED** [review]

**Comment by benthecarman** [comment]
> It's concerning this didn't cause CI failures

We should probably make a `BitcoindRpcAppConfigTest`

**Review by benthecarman — APPROVED** [review]

**Comment by benthecarman** [comment]
`DLCOracleAppConfig` needs its server conf to have the absolute path as well

**Pull request merged**